File: lib/arena/roster.ts
/**
 * RosterManager — tracks the 5 starters (one per position) + 1 flexible bench
 * slot, positional eligibility, and legal roster completion.
 *
 * Anti-cheese: a player can only occupy a slot they're eligible for (primary or
 * secondary position); the bench accepts anyone. This prevents "five point
 * guards masquerading as a lineup" while still allowing flexible bigs (Jokic at
 * C or PF, Giannis at PF/C/SF, etc.).
 */

import {
  type OwnedPlayer,
  type Position,
  type RosterSlot,
  type RosterState,
  type SeasonPlayer,
  ROSTER_SLOTS,
  POSITIONS,
} from "./types"

export function emptyRoster(): RosterState {
  return {
    slots: {
      PG: null,
      SG: null,
      SF: null,
      PF: null,
      C: null,
      BENCH: null,
    },
  }
}

/** All positions a player can legally play. */
export function eligiblePositions(player: SeasonPlayer): Position[] {
  return [player.primaryPosition, ...player.secondaryPositions]
}

/** Can this player fill this slot? Bench accepts anyone. */
export function canFillSlot(player: SeasonPlayer, slot: RosterSlot): boolean {
  if (slot === "BENCH") return true
  return eligiblePositions(player).includes(slot)
}

export function rosterCount(roster: RosterState): number {
  return ROSTER_SLOTS.filter((s) => roster.slots[s] !== null).length
}

export function isRosterComplete(roster: RosterState): boolean {
  return ROSTER_SLOTS.every((s) => roster.slots[s] !== null)
}

/** Slots still empty. */
export function openSlots(roster: RosterState): RosterSlot[] {
  return ROSTER_SLOTS.filter((s) => roster.slots[s] === null)
}

/** Positional slots (excludes bench) that are still empty. */
export function openStarterSlots(roster: RosterState): Position[] {
  return POSITIONS.filter((p) => roster.slots[p] === null)
}

export function ownsPlayer(roster: RosterState, playerId: string): boolean {
  return ROSTER_SLOTS.some((s) => roster.slots[s]?.player.id === playerId)
}

/**
 * Choose the best legal slot for a newly-won player, preferring an open starter
 * slot they're eligible for, then bench. Returns null if there's genuinely no
 * legal slot (roster full or no eligible opening).
 *
 * Preference order among eligible open starter slots: the *scarcer* the player's
 * eligibility, the earlier we lock it in — but at assignment time we simply take
 * the first open eligible starter slot (positions are pre-ordered PG→C). This is
 * deterministic and predictable.
 */
export function bestSlotFor(
  roster: RosterState,
  player: SeasonPlayer
): RosterSlot | null {
  // Prefer a starter slot the player is eligible for.
  for (const pos of eligiblePositions(player)) {
    if (roster.slots[pos] === null) return pos
  }
  // Fall back to bench.
  if (roster.slots.BENCH === null) return "BENCH"
  return null
}

/**
 * The slot this player may take UNDER AUCTION RULES.
 *
 * Starters before bench: you cannot spend your single bench slot while any of
 * your five starting positions is still empty. Without this rule a team could
 * blow real money on a luxury bench piece early and then be stranded needing two
 * starters on a near-empty board (and get silently auto-filled with $1 scrubs).
 * It also stops a team whose starters are set from using its flexible bench slot
 * to hoover up the last eligible player at a position the opponent still needs.
 */
export function auctionSlotFor(
  roster: RosterState,
  player: SeasonPlayer
): RosterSlot | null {
  for (const pos of eligiblePositions(player)) {
    if (roster.slots[pos] === null) return pos
  }
  // No open starter slot fits this player — bench is only legal once the five
  // starting spots are all accounted for.
  if (openStarterSlots(roster).length > 0) return null
  return roster.slots.BENCH === null ? "BENCH" : null
}

/** Returns a NEW roster with the player placed, or throws on an illegal add. */
export function placePlayer(
  roster: RosterState,
  player: SeasonPlayer,
  price: number,
  forcedSlot?: RosterSlot
): RosterState {
  if (ownsPlayer(roster, player.id)) {
    throw new Error(`Duplicate player: ${player.name} is already on this roster.`)
  }

  const slot = forcedSlot ?? bestSlotFor(roster, player)
  if (slot === null) {
    throw new Error(`No legal roster slot available for ${player.name}.`)
  }
  if (roster.slots[slot] !== null) {
    throw new Error(`Slot ${slot} is already filled.`)
  }
  if (!canFillSlot(player, slot)) {
    throw new Error(`${player.name} cannot play ${slot}.`)
  }

  const owned: OwnedPlayer = { player, price, slot }
  return {
    slots: { ...roster.slots, [slot]: owned },
  }
}

/**
 * Whether adding `player` is even *possible* given open slots. Used to gate
 * bidding: never let someone win a player they can't legally roster.
 */
export function canAddPlayer(roster: RosterState, player: SeasonPlayer): boolean {
  if (ownsPlayer(roster, player.id)) return false
  return auctionSlotFor(roster, player) !== null
}

/**
 * Permissive variant for the end-of-auction emergency auto-fill, which must be
 * able to place ANY legal player (including into the bench) to guarantee the
 * game always reaches a complete, playable roster.
 */
export function canForceAddPlayer(roster: RosterState, player: SeasonPlayer): boolean {
  if (ownsPlayer(roster, player.id)) return false
  return bestSlotFor(roster, player) !== null
}

/**
 * Whether the roster can still be *legally completed* if this player is added.
 * Checks that every remaining open starter slot could still be filled by some
 * eligible player from `remainingPool` (a feasibility check to avoid painting
 * yourself into a corner, e.g. taking a C into the last PF slot when you still
 * need a real C and none remain).
 *
 * This is a soft check surfaced as a WARNING in the UI, not a hard block —
 * except when it makes completion provably impossible.
 */
export function completionFeasible(
  roster: RosterState,
  player: SeasonPlayer,
  slot: RosterSlot,
  remainingPool: SeasonPlayer[]
): boolean {
  // Simulate the placement.
  const after = placePlayer(roster, player, 0, slot)
  const needed = openStarterSlots(after)
  if (needed.length === 0) return true

  // Bipartite feasibility (Hall's condition, greedy). For each needed position,
  // is there at least one distinct eligible player left in the pool?
  const pool = remainingPool.filter((p) => p.id !== player.id)
  const used = new Set<string>()
  for (const pos of needed) {
    const candidate = pool.find(
      (p) => !used.has(p.id) && eligiblePositions(p).includes(pos)
    )
    if (!candidate) return false
    used.add(candidate.id)
  }
  return true
}

/** Starters in position order, then bench. Handy for lineup display / sim. */
export function orderedRoster(roster: RosterState): OwnedPlayer[] {
  return ROSTER_SLOTS.map((s) => roster.slots[s]).filter(
    (x): x is OwnedPlayer => x !== null
  )
}

export function starters(roster: RosterState): OwnedPlayer[] {
  return POSITIONS.map((p) => roster.slots[p]).filter(
    (x): x is OwnedPlayer => x !== null
  )
}

export function bench(roster: RosterState): OwnedPlayer | null {
  return roster.slots.BENCH
}

File: auction_ai/simulation/roster.py
"""Roster state + slot rules — faithful port of lib/arena/roster.ts.

This is where player uniqueness and the starters-before-bench rule live. The
learning system may NEVER override these; they are hard auction constraints.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional

from .player_model import POSITIONS, ROSTER_SLOTS, Player


@dataclass
class OwnedPlayer:
    player: Player
    price: int
    slot: str


@dataclass
class Roster:
    """slots maps each RosterSlot → OwnedPlayer | None."""

    slots: Dict[str, Optional[OwnedPlayer]] = field(
        default_factory=lambda: {s: None for s in ROSTER_SLOTS}
    )

    def clone(self) -> "Roster":
        return Roster(slots=dict(self.slots))


def empty_roster() -> Roster:
    return Roster()


def can_fill_slot(player: Player, slot: str) -> bool:
    if slot == "BENCH":
        return True
    return slot in player.eligible_positions()


def roster_count(roster: Roster) -> int:
    return sum(1 for s in ROSTER_SLOTS if roster.slots[s] is not None)


def is_roster_complete(roster: Roster) -> bool:
    return all(roster.slots[s] is not None for s in ROSTER_SLOTS)


def open_slots(roster: Roster) -> List[str]:
    return [s for s in ROSTER_SLOTS if roster.slots[s] is None]


def open_starter_slots(roster: Roster) -> List[str]:
    return [p for p in POSITIONS if roster.slots[p] is None]


def owns_player(roster: Roster, player_id: str) -> bool:
    return any(
        roster.slots[s] is not None and roster.slots[s].player.id == player_id
        for s in ROSTER_SLOTS
    )


def owned_players(roster: Roster) -> List[Player]:
    return [o.player for o in roster.slots.values() if o is not None]


def best_slot_for(roster: Roster, player: Player) -> Optional[str]:
    for pos in player.eligible_positions():
        if roster.slots[pos] is None:
            return pos
    if roster.slots["BENCH"] is None:
        return "BENCH"
    return None


def auction_slot_for(roster: Roster, player: Player) -> Optional[str]:
    """Starters-before-bench: the bench is only legal once all five starter
    positions are accounted for. Port of auctionSlotFor()."""
    for pos in player.eligible_positions():
        if roster.slots[pos] is None:
            return pos
    if len(open_starter_slots(roster)) > 0:
        return None
    return "BENCH" if roster.slots["BENCH"] is None else None


def place_player(
    roster: Roster, player: Player, price: int, forced_slot: Optional[str] = None
) -> Roster:
    """Return a NEW roster with the player placed, or raise on illegal add."""
    if owns_player(roster, player.id):
        raise ValueError(f"Duplicate player: {player.name} already on roster.")
    slot = forced_slot if forced_slot is not None else best_slot_for(roster, player)
    if slot is None:
        raise ValueError(f"No legal roster slot for {player.name}.")
    if roster.slots[slot] is not None:
        raise ValueError(f"Slot {slot} already filled.")
    if not can_fill_slot(player, slot):
        raise ValueError(f"{player.name} cannot play {slot}.")
    new = roster.clone()
    new.slots[slot] = OwnedPlayer(player=player, price=price, slot=slot)
    return new


def can_add_player(roster: Roster, player: Player) -> bool:
    if owns_player(roster, player.id):
        return False
    return auction_slot_for(roster, player) is not None


def can_force_add_player(roster: Roster, player: Player) -> bool:
    if owns_player(roster, player.id):
        return False
    return best_slot_for(roster, player) is not None


def ordered_roster(roster: Roster) -> List[OwnedPlayer]:
    return [roster.slots[s] for s in ROSTER_SLOTS if roster.slots[s] is not None]


def starters(roster: Roster) -> List[OwnedPlayer]:
    return [roster.slots[p] for p in POSITIONS if roster.slots[p] is not None]
